rapids-cmake/test/detail/generate_resource_spec/CMakeLists.txt

```diff
@@ -0,0 +1,34 @@
+#=============================================================================
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+cmake_minimum_required(VERSION 3.23.1)
+project(generate_resource_spec ${lang})
+
+set(CMAKE_CUDA_ARCHITECTURES all)
+
+add_executable(generate_ctest_json generate_resource_spec.cpp)
+if(cuda_toolkit)
+  find_package(CUDAToolkit REQUIRED QUIET)
+  target_link_libraries(generate_ctest_json PRIVATE CUDA::cudart_static)
+  target_compile_definitions(generate_ctest_json PRIVATE HAVE_CUDA)
+endif()
+set_property(SOURCE generate_resource_spec.cpp PROPERTY LANGUAGE ${lang})
+
+set_property(TARGET generate_ctest_json PROPERTY CXX_COMPILER_LAUNCHER)
+set_property(TARGET generate_ctest_json PROPERTY CUDA_COMPILER_LAUNCHER)
```

Comment:
I'm disabling `sccache` here because I determined through extensive debugging that `sccache` is providing a binary that isn't built with `HAVE_CUDA`, even though we're passing `-DHAVE_CUDA` on the command line. Any idea why that is?

Reply:
That is not good at all. Is this happening with a response file that is holding the `-DHAVE_CUDA` or is that just directly on the compile line?

Reply:
I'm reproducing this locally, and just encountered an almost identical issue. A different test was failing because it was trying to link an object file that was apparently built with `-DHAVE_CUDA` without linking `cudart`. The immediately preceding compile line called `sccache` without `-DHAVE_CUDA`. Once I removed `~/.cache/sccache`, this build worked fine, confirming that `sccache` was pulling the wrong object file.

Reply:
I have the smoking gun:

```
[2024-02-08T19:57:00Z DEBUG sccache::server] handle_client: compile
[2024-02-08T19:57:00Z DEBUG sccache::compiler::compiler] Found g++
[2024-02-08T19:57:00Z DEBUG sccache::server] check_compiler: Supported compiler
[2024-02-08T19:57:00Z DEBUG sccache::server] parse_arguments: Ok: ["-DHAVE_CUDA", "-x", "c++", "-fvisibility-inlines-hidden", "-std=c++17", "-fmessage-length=0", "-march=nocona", "-mtune=haswell", "-ftree-vectorize", "-fPIC", "-fstack-protector-strong", "-fno-plt", "-O2", "-ffunction-sections", "-pipe", "-isystem", "/opt/conda/envs/test/include", "-isystem", "/usr/local/cuda/include", "-o", "CMakeFiles/generate_ctest_json.dir/generate_resource_spec.cpp.o", "-c", "/root/src/rapids-cmake/test/detail/generate_resource_spec/generate_resource_spec.cpp"]
[2024-02-08T19:57:00Z DEBUG sccache::compiler::compiler] [generate_resource_spec.cpp.o]: get_cached_or_compile: ["-DHAVE_CUDA", "-x", "c++", "-fvisibility-inlines-hidden", "-std=c++17", "-fmessage-length=0", "-march=nocona", "-mtune=haswell", "-ftree-vectorize", "-fPIC", "-fstack-protector-strong", "-fno-plt", "-O2", "-ffunction-sections", "-pipe", "-isystem", "/opt/conda/envs/test/include", "-isystem", "/usr/local/cuda/include", "-o", "CMakeFiles/generate_ctest_json.dir/generate_resource_spec.cpp.o", "-c", "/root/src/rapids-cmake/test/detail/generate_resource_spec/generate_resource_spec.cpp"]
[2024-02-08T19:57:00Z DEBUG sccache::compiler::c] Preprocessor cache hit: 3e4bc0b67e125bb2248d3e0882c7b0b71b47929ff11fedcd35fbd9719c3384ab
[2024-02-08T19:57:00Z DEBUG sccache::compiler::compiler] [generate_resource_spec.cpp.o]: generate_hash_key took 0.015 s
[2024-02-08T19:57:00Z DEBUG sccache::compiler::compiler] [generate_resource_spec.cpp.o]: Hash key: c4ca345385664a53a11bf15dccc49cf63df600bcb35cad0feb97830defd9aee6
[2024-02-08T19:57:00Z DEBUG sccache::compiler::compiler] [generate_resource_spec.cpp.o]: Cache hit in 0.018 s
[2024-02-08T19:57:00Z DEBUG sccache::server] compile result: cache hit
[2024-02-08T20:09:46Z DEBUG sccache::server] handle_client: compile
[2024-02-08T20:09:46Z DEBUG sccache::server] check_compiler: Supported compiler
[2024-02-08T20:09:46Z DEBUG sccache::server] parse_arguments: Ok: ["-x", "c++", "-fvisibility-inlines-hidden", "-std=c++17", "-fmessage-length=0", "-march=nocona", "-mtune=haswell", "-ftree-vectorize", "-fPIC", "-fstack-protector-strong", "-fno-plt", "-O2", "-ffunction-sections", "-pipe", "-isystem", "/opt/conda/envs/test/include", "-isystem", "/usr/local/cuda/include", "-o", "CMakeFiles/generate_ctest_json.dir/generate_resource_spec.cpp.o", "-c", "/root/src/rapids-cmake/test/detail/generate_resource_spec/generate_resource_spec.cpp"]
[2024-02-08T20:09:46Z DEBUG sccache::compiler::compiler] [generate_resource_spec.cpp.o]: get_cached_or_compile: ["-x", "c++", "-fvisibility-inlines-hidden", "-std=c++17", "-fmessage-length=0", "-march=nocona", "-mtune=haswell", "-ftree-vectorize", "-fPIC", "-fstack-protector-strong", "-fno-plt", "-O2", "-ffunction-sections", "-pipe", "-isystem", "/opt/conda/envs/test/include", "-isystem", "/usr/local/cuda/include", "-o", "CMakeFiles/generate_ctest_json.dir/generate_resource_spec.cpp.o", "-c", "/root/src/rapids-cmake/test/detail/generate_resource_spec/generate_resource_spec.cpp"]
[2024-02-08T20:09:46Z DEBUG sccache::compiler::c] Preprocessor cache hit: 3e4bc0b67e125bb2248d3e0882c7b0b71b47929ff11fedcd35fbd9719c3384ab
[2024-02-08T20:09:46Z DEBUG sccache::compiler::compiler] [generate_resource_spec.cpp.o]: generate_hash_key took 0.003 s
[2024-02-08T20:09:46Z DEBUG sccache::compiler::compiler] [generate_resource_spec.cpp.o]: Hash key: c4ca345385664a53a11bf15dccc49cf63df600bcb35cad0feb97830defd9aee6
[2024-02-08T20:09:46Z DEBUG sccache::compiler::compiler] [generate_resource_spec.cpp.o]: Cache hit in 0.000 s
[2024-02-08T20:09:46Z DEBUG sccache::server] compile result: cache hit
```

Reply:
I have a script that reproduces the issue in a minimal way:

```
# docker run -it --rm rapidsai/ci-conda:cuda11.4.3-centos7-py3.9

set -xe

cat > /root/main.cpp <<'EOF'
/*
 * Copyright (c) 2022-2024, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifdef HAVE_CUDA
#include <cuda_runtime_api.h>
#endif

#include <iostream>
#include <string>
#include <vector>

struct version {
  version() : json_major(1), json_minor(0) {}
  int json_major;
  int json_minor;
};

struct gpu {
  gpu(int i) : id(i), memory(0), slots(0){};
  gpu(int i, size_t mem) : id(i), memory(mem), slots(100) {}
  int id;
  size_t memory;
  int slots;
};

// A hard-coded JSON printer that generates a ctest resource-specification file:
// https://cmake.org/cmake/help/latest/manual/ctest.1.html#resource-specification-file
void to_json(std::ostream& buffer, version const& v)
{
  buffer << "\"version\": {\"major\": " << v.json_major << ", \"minor\": " << v.json_minor << "}";
}
void to_json(std::ostream& buffer, gpu const& g)
{
  buffer << "\t\t{\"id\": \"" << g.id << "\", \"slots\": " << g.slots << "}";
}

int main()
{
  std::vector<gpu> gpus;
  int nDevices = 0;

#ifdef HAVE_CUDA
  cudaGetDeviceCount(&nDevices);
  if (nDevices == 0) {
    gpus.push_back(gpu(0));
  } else {
    for (int i = 0; i < nDevices; ++i) {
      cudaDeviceProp prop;
      cudaGetDeviceProperties(&prop, i);
      gpus.push_back(gpu(i, prop.totalGlobalMem));
    }
  }
#else
  gpus.push_back(gpu(0));
#endif

  version v;
  std::cout << "{\n";
  to_json(std::cout, v);
  std::cout << ",\n";
  std::cout << "\"local\": [{\n";
  std::cout << "\t\"gpus\": [\n";
  for (int i = 0; i < gpus.size(); ++i) {
    to_json(std::cout, gpus[i]);
    if (i != (gpus.size() - 1)) { std::cout << ","; }
    std::cout << "\n";
  }
  std::cout << "\t]\n";
  std::cout << "}]\n";
  std::cout << "}" << std::endl;
  return 0;
}
EOF

cat > /root/env.yaml <<'EOF'
# This file is generated by `rapids-dependency-file-generator`.
# To make changes, edit dependencies.yaml and run `rapids-dependency-file-generator`.
channels:
- rapidsai
- conda-forge
dependencies:
- c-compiler
- cmake>=3.23.1,!=3.25.0
- cuda-version=11.4
- cudatoolkit
- cxx-compiler
- cython>=0.29,<0.30
- fmt==9.1.0
- gcc<11.0.0
- libpng
- make
- ninja
- nvcc_linux-64=11.4
- pip
- python>=3.9,<3.11
- sphinx
- sphinx-copybutton
- sphinx_rtd_theme
- sysroot_linux-64==2.17
- zlib
- pip:
  - sphinxcontrib-moderncmakedomain
name: test_cuda-114_arch-x86_64

EOF

. /opt/conda/etc/profile.d/conda.sh

rapids-mamba-retry env create --force -f /root/env.yaml -n test

conda activate test

unset SCCACHE_BUCKET

SCCACHE_ERROR_LOG=/root/sccache.log SCCACHE_LOG=debug sccache --start-server

sccache --version

sccache /opt/conda/envs/test/bin/x86_64-conda-linux-gnu-c++   -x c++ -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/test/include -isystem /usr/local/cuda/include -o /root/main.nocuda.cpp.o -c /root/main.cpp

sccache /opt/conda/envs/test/bin/x86_64-conda-linux-gnu-c++ -DHAVE_CUDA  -x c++ -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/test/include -isystem /usr/local/cuda/include -o /root/main.cuda.cpp.o -c /root/main.cpp

cat /root/sccache.log
```

Reply:
We Traced this back to the sccache issue: https://github.com/mozilla/sccache/issues/2038

Which is fixed in 0.7.6

Reply:
Now that we've identified `sccache` as the issue, and will soon have a fix, I've reinstated `sccache` here. Once we respin the CI images, we can rerun this job and it will work.